### ISSUE TITLE
[mu4e bug] mu loses track of messages marked as (un)read in mu4e

### ISSUE THREAD

**Issue by Alexander-Miller** [djcb/mu#2741]
**Describe the bug**

A message which goes from read to unread (and vice versa) in mu4e can not be found by `mu find`. Changing the read state to its original position makes the message findable again. The message can also be found after closing Emacs.

**How to Reproduce**

1: Start with a single unread message in mu4e that can be found in mu:

![1](https://github.com/user-attachments/assets/694f7fbb-1117-4f52-9806-615986b463d6)

2: Mark the message as read with `mu4e-headers-mark-for-read`. mu has now lost track of this message:

![2](https://github.com/user-attachments/assets/f7d9fa70-f934-41c0-8dbd-5c52705499e1)

3: Revert the change with `mu4e-view-mark-for-unread`. The message now appears in mu again:
![3](https://github.com/user-attachments/assets/2db0ccd6-edfa-4247-9ef2-ae784687c0d7)

mu will also be able to find the message if you quit Emacs after step 2.

According to the documentation there might be differences between mu and mu4e due to sorting, limits and the include-related flag. However inconsistencies when reading individual messages seems like a bug to me.

For context: I am building a treemacs extension that gives mu4e a thunderbird-inspired sidebar, where I want to include at least [a count of all unread emails](https://imgur.com/a/18dNVd2).  From what I saw in the mu4e codebase its communication with the server is a closed loop, so I need to use the mu cli to get those mail counts.

**Environment**

- EndeavourOS
- Emacs 29.4 homebrew config
- mu 1.12.6

**Checklist**

- [x] you are running either an 1.10.x/1.12.x release or `master` (otherwise please upgrade)
- [x] you can reproduce the problem without 3rd party extensions (including Doom/Evil, various extensions etc.)
- [x] you have read all of the above

**Comment by djcb** [comment]
For performance reasons, the mu doesn't write the database to disk for each operation; so you can't really expect to immediately get the same results from the command-line.  Iie. should add an item for that to the FAQ, as it has come up before.

Anyway, I'd avoid running command-line queries, it's a bit inefficient. But perhaps mu4e could have some API to get you the information directly; it'd be limited to the information mu4e already  has available though.

**Comment by Alexander-Miller** [comment]
> Anyway, I'd avoid running command-line queries, it's a bit inefficient.

Doing it about 30 times in a row right now (by shelling out to python first so Emacs isn't blocked) and that works pretty well, at least for my mailbox of <1000 emails :shrug: . It's really only the staleness of the data that's an issue.

> But perhaps mu4e could have some API to get you the information directly; it'd be limited to the information mu4e already has available though.

That'd be great. The basic ingredients are already present since mu4e will already show the same numbers I also want for the content of `mu4e-maildir-shortcuts`. I'd be willing to come up with a PR if you advise me on an implementation that doesn't need to touch the C++ parts.

**Comment by djcb** [comment]
Updated the FAQ.

As for using the information in mu4e directly, the available information is in `mu4e-query-items`. I've extended its documentation a bit, hope this is sufficient for your use-case!

**Comment by Alexander-Miller** [comment]
> hope this is sufficient for your use-case!

Unfortunately not. The info I am looking for is numbers for every individual maildir, or at least for every email address. `mu4e-query-items` gives info based on `mu4e-maildir-shortcuts`. I, for example, don't use that feature in my own config, so I get nothing. Trying to get info that way means mandating the content of `mu4e-maildir-shortcuts` - bad UX and a misappropriation of a related feature. let-binding it won't work either because asking the server is asynchronous.

> Note that quitting {mu4e} or invoking {mu4e-update-mail-and-index}
and {mu4e-update-index} trigger committing to disk as a side-effect.

The update-index commands don't do that for me, neither did a write happen 'eventually' and the above example stuck around for days. Only quitting mu triggered the write (which I only found out about today because I put packages like mu into their own eyebrowse desktop, so I never really quit them).

**Comment by djcb** [comment]
Ah, then I misunderstood your earlier comment, "The basic ingredients are already present since mu4e will already show the same numbers I also want". Anyway, as I said, you only have access to the information that mu4e already has, if you want something else you're out of luck, unfortunately.

Getting the count/unread counts for _all_ maildirs can be costly, though it's not too bad if you don't have too many maildirs. 

As for the time until mu4e writes thing to disk, for me that works just as I mentioned, I wonder why you see something else. 

So basically, in mu4e:
1. mark message as unread
2. M-x mu4e-update-index

after 1. you cannot find the message with `mu find`? And not after 2. either?

**Comment by Alexander-Miller** [comment]
> after 1. you cannot find the message with mu find? And not after 2. either?

Correct. Here's a video what it looks like:

https://github.com/user-attachments/assets/18f5d647-7dfd-4ac5-bd4d-570c56cb7f7e

First mu can find the email. Then I mark is as unread and mu loses it. Running mu-update commands does not help. The email is found again after I close Emacs.
